### DISCUSSION TITLE
vweb refactor proposal

### DISCUSSION THREAD

**Discussion by mcastorina** [vlang/v#15187] [Ideas]
## Intention
The vweb module will be refactored to have a structure suitable for parallelization.  As a community it would be best to agree on the design collectively.

The intent of this issue is to settle on a design for the refactored vweb module.

## Proposal
Use `vweb.Context` for request specific context, and use the user supplied App struct strictly for state management. Methods defined on `App` will be declared using `shared` to enable developers to use `lock` and `rlock` as necessary.

Additionally, `vweb.run` will take two arguments: an initialized generic type and a `vweb.Config` struct. This allows developers to initialize anything they may need before server start.

### Example

```v
import vweb

struct App {
mut:
	cnt int
}

fn main() {
	conf := vweb.Config{
		port: 8082
		static_directory: '.'
		static_path: '/static'
	}
	vweb.run(App{cnt: 100}, conf)
}

['/users/:user']
fn (shared app App) user_endpoint(mut c vweb.Context, user string) vweb.Result {
    rlock {
        return c.json('{"$user": $app.cnt}')
    }
}

fn (shared app App) index(mut c vweb.Context) vweb.Result {
    lock {
        app.cnt++
    }
	// These values are available in the index.html template
	show := true
	hello := 'Hello world from vweb'
	numbers := [1, 2, 3]
	c.enable_chunked_transfer(40)
	return $vweb.html()
}
```

## Technical Details
I haven't looked too deep into it, but I think this is not easily possible. I tried and failed making a POC passing a `shared` struct to a generic function, or initializing a shared generic type. So here is my question: will it be possible to do these things with V or future versions of V?

**Comment by spytheman** [comment]
I think this ` lock {` should be ` lock app {` , also `rlock app {`

**Comment by spytheman** [comment]
I think the proposed program is nice and clear, and agree with the design changes.

@medvednikov what do you think?

**Comment by spytheman** [comment]
> passing a shared struct to a generic function, or initializing a shared generic type. 

That needs further investigation, but I do not see a fundamental problem for now; it may be just a bug that needs fixing.
@UweKrueger what do you think?

**Comment by UweKrueger** [comment]
> > passing a shared struct to a generic function, or initializing a shared generic type.
> 
> ... it may be just a bug that needs fixing.
> @UweKrueger what do you think?

I aggree. This should be made possible. I'll work on it.

**Comment by mcastorina** [comment]
@Zerotask it's the best way I can think of to get request specific context to the user's callback function. [Gin](https://github.com/gin-gonic/gin) uses a similar method.

We can't easily embed request specific context into the `App` structure because each concurrent request has unique data. If they all share the single `Context` like they do now, there will be race conditions and unexpected behavior with concurrent calls.

I'm open to suggestions if you have any.

**Comment by medvednikov** [comment]
@mcastorina the main idea behind vweb is that there are "global" objects that must shared across requests like DB connections. One connection across all clients.

And of course there is data unique to each user, like current user object, ip, etc.

Right now they are all in one struct, which is a mess, and also means the developer has to manually reset user fields in `.init()` :)

**Comment by mcastorina** [comment]
With that in mind, I think my proposal will fit vweb's main idea:

- User provided struct `App` is the "global" that objects share across requests
- The initialized `App` is passed directly to `vweb.run` (this is equivalent to `init_once()`)
- `Context` struct passed to the callback is used for request data / sending response
- `shared` methods for `App` to prevent concurrency issues

Am I missing anything? What are your thoughts on this proposal @medvednikov ?

**Comment by JalonSolov** [comment]
So `Context` is where you would keep the data unique to each user?

**Comment by mcastorina** [comment]
I would say unique to each request, not user. Things like

- HTTP request struct (headers, body, etc.)
- query parameters
- ~cookies~ maybe cookies doesn't fit here.. need to think about it more
- HTTP response struct (headers, body, etc.)

If the app is operating on users, then in the callback you would have to query the DB for the user struct through `App`.

In other words, `Context` is a transient struct. It's created per request and is destroyed once the response is sent.

**Comment by medvednikov** [comment]
I think it sounds good.

One other major thing vweb is missing right now, is imitating controllers functionality in MVC frameworks. For example, having an `Admin` struct with access to all fields, but also taking care of authorization, and handling each url starting with `/admin/`.

**Comment by mcastorina** [comment]
I have been thinking a little about this.. so far I've come up with middlewares that can be added to the `Config` struct.

```v
fn main() {
	conf := vweb.Config{
		port: 8082
		middlewares: {
			'/admin': [auth_admin]
		}
	}
	vweb.run(App{cnt: 100}, conf)
}

[middleware]
fn (shared app App) auth_admin(mut c vweb.Context) vweb.Result {
	auth_header := c.headers.get(.authorization)
	// ...
	if is_authorized {
		return c.next()
	}
	return c.text('unauthorized')
}

fn (shared app App) admin(mut c vweb.Context) vweb.Result {
	return c.text('welcome authorized admin!')
}
```

The idea isn't fully fleshed out, but I think having a central place to register a middleware with a path prefix is safer than adding the middleware to each callback.

**Comment by JalonSolov** [comment]
Hmm... seems to me that we have 3 sets of data:

* global to the app.  Applies to all instances, requests, etc.
* Per user of the app.
* Per individual request.

As far as multiple middleware with a single attribute... I think that will become too unwieldy, too quickly, especially if you want to allow middleware plugins.

**Comment by mcastorina** [comment]
Another technical limitation that needs to be implemented for this is passing structs with compile-time reflection. I opened a feature request here: https://github.com/vlang/v/issues/9103

This is needed for passing the `vweb.Context` struct to the user defined methods.

**Comment by medvednikov** [comment]
@emily33901 did you figure out the reason io.BufferedReader is slow?

Perhaps we can go back to direct sockets, without interfaces?

**Comment by mcastorina** [comment]
@Zerotask could you share the program you used to test? The vweb example is giving me ~1.5ms response times

**Comment by JalonSolov** [comment]
A lot of good suggestions... most of which belong in separate issues.  This one is about refactoring vweb, not about adding new functionality.  Let's try to get one thing done at a time, otherwise the PR gets huge, messy, and hard to review.

**Comment by JalonSolov** [comment]
If we do that, we need to change the title of this issue.  I am all for getting vweb to be the best we can get it, I just hate to wade through fifteen unrelated changes in the same PR.

PRs should always be as focused on a single change as they can be.  Do extra work in separate PRs.

**Comment by fnetX** [comment]
Sorry for stepping in there a bit late, but I'm confused by the new syntax, it looks a bit off to put `mut c vweb.Context` into every method.
Isn't it somehow possible to use one struct that references the original app (so it's stateful) and embeds a new context for each request? Thinking of creating a struct that embeds the existing app state somehow ... just thinking aloud, it would be much more convenient + would not be a breaking change.

**Comment by medvednikov** [comment]
@fnetX I was thinking the same.

It feels wasteful and repetitive adding `mut c vweb.Context` to each method.

**Comment by mcastorina** [comment]
Hmm.. that's what I proposed simply because that is what I'm used to (with Gin in the Go world).

I'm not sure how else to differentiate unique requests if `App` is shared between them. The best I can come up with is instead of passing `mut c vweb.Context` we pass a `ctx_id int` that can be looked up from `App` if needed.

Another option is to create a new instance of `App` for each request, but then we need to figure out global state.

**Comment by mcastorina** [comment]
Or `memcpy` `App` to embed the `Context`? But that seems wasteful, especially for large apps.

**Comment by emily33901** [comment]
Hi - sorry I haven't gotten to this thread sooner I'll read over it in the coming days

**Comment by fnetX** [comment]
I'm not yet that deep in the language, but isn't it possible to create a Context struct that holds a reference to the mutable app struct, but is created anew for every request?

something like this, but in the proper done way :joy: 
~~~
struct App {
mut:
    data string
}

struct Context {
    &App
    vweb.Context
}
~~~
and
~~~
mut app := App {} // create app only once at startup
for request ... {
    // somehow combine a mutable ref to the app with the request context
    // the latter one being created for every request, the app being embedded
    handle_conn(Context{&app, request})
}
~~~

**Comment by mcastorina** [comment]
```v
import vweb

struct App {
	id int
}

struct AppContext {
	app &App // pointer to global state
	vweb.Context // embed vweb's Context
}

// handles are defined on the combined struct
fn (mut app AppContext) index() vweb.Result {
	mut id := 0
	rlock app.app {
		id = app.app.id
	}

	return app.json('{"id": $id}')
}
```

Something like this could work, but I see a few problems with it (imo):
1. It would require the user of vweb to create two structs
2. I'm not sure if you can `lock` a pointer / use a `shared` object in a struct
3. (My personal opinion) it blurs the lines between global data and request specific data too much

**Comment by medvednikov** [comment]
How is this problem (separating request local data vs global data like db connections) handled in other frameworks like Rails, .NET MVC, Django, Flask etc?

Do they even allow global data?

**Comment by dumblob** [comment]
Can't speak for Rails, .NET MVT, Django. But in Flask they do allow global data, but it's utterly slow and not recommended to use (though it's safe). Note, that in older versions of Flask, the "global" object was there, but wasn't behaving correctly in multithreaded/multiprocess environments (this weird behavior was documented, but people kept using it incorrectly, so they made it safe but at the expense of even more slower access).

I personally don't think global state shall be made accessible this way. IMHO it should be more explicit without "blurring the line" as @mcastorina pointed out above. Making it explicit has also (potentially significant) performance benefits and seems much more future-proof.

**Comment by fnetX** [comment]
I'd also think that it's good if global data is explicit. I like [this syntax](https://github.com/vlang/v/issues/9081#issuecomment-815803300) and it's flexible (you could reference more than app, but for example have another global object, right? (instead of AppContext I'd just use Context probably)
A good thing would be to only require this definition if you want to have global state and leave it out if you don't.

**Comment by medvednikov** [comment]
What about `[global]` ?

```
struct App {
  	vweb.Context
	id int
	db &sql.DB [global]
}

```

**Comment by dumblob** [comment]
> What about `[global]` ?

What would be the precise semantics (regarding per-request, per-thread, per-process, per-DB file, per-DB name, etc.)?

**Comment by medvednikov** [comment]
per server

**Comment by medvednikov** [comment]
can also use a more clear name `[server_var]` or something

**Comment by dumblob** [comment]
Hm, so it's vweb-specific and can't be used anywhere else?

**Comment by fnetX** [comment]
> can also use a more clear name [server_var] or something

Please no more vweb compiler magic. Introducing general features like global parts of structs sounds fine, if it is not achievable with the current implementation, but please nothing too specific.

**Comment by medvednikov** [comment]
[attributes] can be used by anything, not just vweb.

**Comment by fnetX** [comment]
Ah, I'm sorry, you mean to use the existing compiler features and reference the global struct as discussed earlier, but use an attribute for internal vweb handling? Sounds good to me.

**Comment by JalonSolov** [comment]
This _particular_ attribute (if `[global]` is used) should probably be vweb specific, otherwise people will try to use it to get around the `__global` + `--enable-globals` restrictions.

In vweb, it means something different, anyway.  So `[server_var]` or something along those lines makes more sense.

**Comment by mcastorina** [comment]
I have a few questions about this:

```v
struct App {
  	vweb.Context
	id int
	db &sql.DB [global]
}
```

Personally, I don't like this approach, but if we do go forward with it..

1. What is `id`? It's not "global" attribute, or part of `vweb.Context`. It would get reset to 0 every time.
2. How will this work on a technical level? I imagine a new `App` struct being created, memcpy the `[global]` flags / set the `Context` data.
3. Will users be responsible for thread safe access to "global" attributes?

Thinking about 1 and 2.. Why even bother with a `[global]` attribute anyway? Non `[global]` variables would be useless being reset to their default. Might as well clone the entire `App` every time.

**Comment by medvednikov** [comment]
`id` is request local data, like `user_name` for example, it's ok for it to be reset. The developer will set it to the right value in `App.init()`.

Yes, global data has to be copied for each new `App` created for each request.

No, thread safe access will be guaranteed by vweb.

**Comment by medvednikov** [comment]
I agree with @JalonSolov, "global" is a bad name. `server_var` is better.

**Comment by mcastorina** [comment]
I'm still not convinced this syntax is better, only to save a few keystrokes. Here are the pros / cons as I see it:

1. Pro: do not have to use `mut c vweb.Context` in all handlers
2. Pro: Adds a place for request local data
3. Con: developers need to be educated about the `[server_var]` attribute
  a. education cannot happen through compiler warnings / errors (i.e. not easily discoverable)
4. Con: overhead of copying `App` -- server performance suffers the larger `App` is

**Comment by medvednikov** [comment]
`App` is created for each request, what do you mean by copying? Only the shared `[server_var]` reference is copied.

**Comment by medvednikov** [comment]
Most web applications probably create new DB connections on each request (sometimes via a pool), so vweb's `[server_var]` would be just a low level optimization for most cases.

**Comment by mcastorina** [comment]
`App` being created for each request requires a malloc, with additional time spent copying `[server_var]` attributes to the new object.

This overhead is not needed in the original proposal -- only `Context` objects would be created per request. That is why I see it as a con.

**Comment by JalonSolov** [comment]
"Only the shared `[server_var]` **_reference_** is copied."

Unless you had a few thousand of these, we're not talking a lot of time.  And if they're kept together, it could be a single memcpy of the fields, just copying a set of pointers.

**Comment by mcastorina** [comment]
In *addition* to the newly instantiated `App` struct. I agree, the overhead of copying the `[server_var]` attributes may be negligible, but a `malloc` is not.

**Comment by medvednikov** [comment]
@mcastorina but user data has to be allocated for each request anyway, so I don't see any extra overhead.

**Comment by dumblob** [comment]
> @mcastorina but user data has to be allocated for each request anyway, so I don't see any extra overhead.

They can be allocated in thread local storage (which in turn is a "preallocated" space at thread creation time) and thus the access is much faster than `malloc()` (but still slower than access to stack data). So I still agree with @mcastorina that forcing `malloc()` per request **by design** is a wrong choice for *V* built-in web server.

**Comment by medvednikov** [comment]
The same can be done with this approach, there's no difference.

**Comment by dumblob** [comment]
> The same can be done with this approach, there's no difference.

Then I don't understand the whole concept with `[server_var]`. Is it really just about (a) versus (b) as follows?

a) `malloc()`/`tls_alloc()` of `App` per request with subsequent filling with data (including `[server_var]` fields) while having to access the data from a request handler method with `app.

b) `malloc()`/`tls_alloc()` of `App` per thread (i.e. once - namely at the init of the thread pool) and per request filling with data (probably without a reference to `App` because it can be obtained from context) while having to have by-one-argument longer argument list of the request handler method

If it's really just this, then I think the difference in convenience is negligible and the fact that (b) API allows avoiding any `malloc()`/`tls_alloc()` seems a much better proposition. But maybe I'm still missing something.

**Comment by medvednikov** [comment]
I didn't consider a single allocation per thread. But it's doable with both syntaxes.

**Comment by medvednikov** [comment]
By the way, Flask has the same global DB handle approach:

https://github.com/pallets/flask/blob/1.1.2/examples/tutorial/flaskr/db.py

They use `flask.g`, the global application context for that.

https://flask.palletsprojects.com/en/1.1.x/appcontext/

**Comment by medvednikov** [comment]
In general, I think Flask is a good source of inspiration. It's very well designed, and is quite clean and minimalistic.

**Comment by dumblob** [comment]
@medvednikov yes, Flask does that as I wrote above in https://github.com/vlang/v/issues/9081#issuecomment-817650772 . But I'm convinced it's not as clean as it might seem. It supports non-systemic design of systems and is basically useful for debugging, but I wouldn't recommend it for production (this is all from my experience with production-level Flask systems).

**Comment by mcastorina** [comment]
I have been thinking about this more, and I don't understand how mutable variables will be achieved with the `[server_var]` syntax. 

We can copy the old values to the new `App` struct, but how will the updated values be set for the next requests? Read the `[server_var]` fields at the end of the handler and update the "reference" object? That will introduce race conditions unless the entire handler is locked (which takes us back to single threading).

Personally, I prefer putting `Context` as the first argument to the handler. Request local data can be supported by changing it to be an interface too. This would allow developers to create their own struct with `vweb.Context` embedded into it.

**Comment by medvednikov** [comment]
@mcastorina good point about race conditions.

Perhaps this should only be allowed for DB handles, since DBs automatically handle those.

And I can also only think of DB handles requiring this global status at all.

**Comment by fnetX** [comment]
> Personally, I prefer putting Context as the first argument to the handler.

Why is there apparently no solution to do the same in one struct? In my understanding, you'd have the global struct + create a new context struct for each request, and have the function operate on the first with passing the context struct.
What's the difference to having a global struct and creating a context struct that refers to this + having the function operate on the context struct directly?

Also, I kinda disapprove to have mutable global data as this leads developers to build stateful apps, which are harder to scale across machines. Only having persistent state (like config and database connections) and everything else in an external tool (like a database or memory cache) makes sense to me. So I don't really get what needs to be changed from the current behaviour, and especially I don't get why I'd have to spell out the Context struct each time I'm creating a vweb function. This just looks unnecessarily ugly to me.

**Comment by mcastorina** [comment]
> What's the difference to having a global struct and creating a context struct that refers to this + having the function operate on the context struct directly?

I'm trying to figure out what this would look like. V cannot define new methods on non-local types, so the user has to define a new `Context` struct (or vweb will have to figure out another way to detect routing / handlers). Then a separate struct will be needed for the "global" state that the `Context` struct references.

```v
struct Handler { vweb.Context }
struct App { id int }

fn main() {
    vweb.run<Handler, App>(8080)
}

fn (mut h Handler) index() vweb.Result {
    mut id := 0
    lock h.app {
        id = h.app.id
        h.app.id++
    }
    return h.text('id: $id')
}
```

I think this looks fine and is technically feasible. I'm not sure how to go about making `App` optional though.

Maybe something like this..
```v
fn main() {
    vweb.run<Handler>(8080)
        .manage<App>(App{})
}
```
or some kind of configuration.

**Comment by medvednikov** [comment]
I've already fixed it on my local branch, will push it today.

**Comment by medvednikov** [comment]
Had to implement `$if x is Interface{}` for this.

**Comment by dumblob** [comment]
>V cannot define new methods on non-local types, so the user has to define a new `Context` struct (or vweb will have to figure out another way to detect routing / handlers).

This pops up here and there sometimes. For those interested there is a discussion about this in https://github.com/vlang/v/issues/2025 (and related https://github.com/vlang/v/commit/895cefe351381955bd38057f581c9bc7a20598a3 and https://github.com/vlang/v/issues/9735 ). I myself am still a bit more inclined to getting this restriction lifted at some point - maybe after *V 1.0*, but let's discuss it in the other thread :wink:.

**Comment by atorr0** [comment]
Hi all,

IMHO I would like to propose using **channels** to delegate to worker threads. I have done some tests and it seems to me fast enough to be considered as an option; something like this (heavily inspired by vweb):

```v
struct TcpConWrapper {
mut:
	conn net.TcpConn
}

pub struct FullServerContext {
mut:
	// sctx  ServerContext
	tcpcw chan TcpConWrapper
}

fn run_fsc(mut fsc FullServerContext) {
	mut l := net.listen_tcp(8080) or { panic('failed to listen') }
	log(0, 'Listening http://localhost:8080 ...')

	for i := 0; i < runtime.nr_jobs(); i++ {
		go sworker(1 + i, fsc)
	}

	for {
		mut conn := l.accept() or { panic('accept() failed') }
		fsc.tcpcw <- TcpConWrapper{
			conn: conn
		}
	}
}

[manualfree]
fn sworker(tid int, fsc FullServerContext) {
	for {
		log(tid, 'Waiting to serve...')
		mut tcpcw := <-fsc.tcpcw
		log(tid, 'Begin service.')

		init_conn(mut tcpcw.conn)
		sworker_main(tid, mut tcpcw)

		tcpcw.conn.close() or { eprintln('error10 closing connection: $err') }
	}
}
```

**Comment by medvednikov** [comment]
i've just pushed a big commit that makes vweb thread safe
it now creates a new app object for each connection and copies the "global" db reference
i decided to limit global fields to avoid race conditions and all other problems
so the global fields that can be shared are defined by vweb now, not by the user

**Comment by mcastorina** [comment]
> so the global fields that can be shared are defined by vweb now, not by the user

This assumes the object `db` points to is thread safe. Technically the user could still create their own struct that implements `DbInterface` that is not thread safe.

---

I have mixed feelings about this direction, so I'll list the pros / cons as I see them right now.

Pros:
1. "one way" philosophy makes development easier
2. prevents transient global state in web apps

Cons:
1. Implicit behavior - your `App` needs to use `db` as the variable name for it to be "global" across requests
2. Cannot have more than one db

**Comment by dumblob** [comment]
Thanks @mcastorina for the quick pros&cons lists. I for myself am least satisfied with the somewhat implicit "nudging" to use just one DB (i.e. to "centralize several DBs under one such point") because I've never yet made any web app which would use just one DB (and I'm talking about both small web apps as well as about big scale web apps).

I'm curious where this approach goes further...

**Comment by medvednikov** [comment]
@mcastorina yes you are right, fortunately this is a temporary limitation.

Right now it looks at the field name, later it will look at DB types, so both of these will be fixed.

**Comment by dumblob** [comment]
>Right now it looks at the field name, later it will look at DB types, so both of these will be fixed.

That gives us a hope :wink:. Will the "DB types" be user-definable? Because I'm usually using own DB proxy APIs.

**Comment by fnetX** [comment]
Since we were talking about those properties earlier, I'd still vote for some `[global]` property, so if I want to include something else I should be able to do so IMO.

**Comment by JalonSolov** [comment]
You should be able to specify the db type with attributes, instead of requiring globals.

**Comment by fnetX** [comment]
I mean, using an attribute to specify which data is copied, the naming of global was kinda bad ... but if it's changing anyway, we don't have to bother with improving the workaround?

**Comment by JalonSolov** [comment]
No, I was thinking more along the lines of
```v
pub struct FullServerContext {
mut:
	// sctx  ServerContext
	tcpcw chan TcpConWrapper

        [db: mysql]
        my_db1 <whatever>

        [db: mysql]
        my_db1 <whatever>
}
```
